Don't require a full layout to query size of SVG renderers.

mattwoodrow opened the pull request:
#### 64a2c5d4f8735c2f7007fce7ada27e2418ed19e9
<pre>
Don&apos;t require a full layout to query size of SVG renderers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291185">https://bugs.webkit.org/show_bug.cgi?id=291185</a>
&lt;<a href="https://rdar.apple.com/148727795">rdar://148727795</a>&gt;

Reviewed by Said Abou-Hallawa.

For SVG content, as long as the target doesn&apos;t need any form of layout, and
ancestors don&apos;t need self or positioned movement layout, then we should be able
to use the existing layout values.

It also updates clientLeft/Top to use this path (like width/height already do),
and adds a few missing ayoutOptions::IgnorePendingStylesheets flags.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetWidth):
(WebCore::Element::offsetHeight):
(WebCore::Element::clientLeft):
(WebCore::Element::clientTop):
(WebCore::Element::clientWidth):
(WebCore::Element::clientHeight):
(WebCore::Element::scrollWidth):
(WebCore::Element::scrollHeight):

Canonical link: <a href="https://commits.webkit.org/293444@main">https://commits.webkit.org/293444@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80bc8d453076fe3bd27cd14a299669a80eca3ccc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18572 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104062 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49526 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100980 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18867 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75314 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32442 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101939 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14332 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55675 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14124 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7315 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48904 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84060 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7389 "Found 2 new test failures: pdf/pdf-in-embed.html pdf/pdf-in-styled-embed.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106430 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26032 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18974 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84277 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26407 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85538 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83778 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21240 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28433 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6102 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19754 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25985 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31171 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25805 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29125 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27379 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3170,6 +3170,14 @@ bool Document::updateLayoutIfDimensionsOutOfDate(Element& element, OptionSet<Dim
         for (CheckedPtr currentRenderer = renderer; currentRenderer && !currentRenderer->isRenderView(); currentRenderer = currentRenderer->container()) {
 
+            if (currentRenderer->isSVGRenderer()) {
+                if (currentRenderer->selfNeedsLayout() || currentRenderer->needsPositionedMovementLayout()) {
```
Should we check `dimensionsCheck` the same way we do below with currentBox:
```
    if (dimensionsCheck.containsAny({ DimensionsCheck::Left, DimensionsCheck::Top }) && (currentRenderer->selfNeedsLayout() || currentRenderer->needsPositionedMovementLayout()))
```

mattwoodrow replied:
Yeah, that's a good idea.

I've refactored things a bit so that checks that are common to both types happen first, and then we do box/svg branch for type specific checks.

shallawa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3170,6 +3170,14 @@ bool Document::updateLayoutIfDimensionsOutOfDate(Element& element, OptionSet<Dim
+                    requireFullLayout = true;
+                    break;
+                }
+                continue;
```
Do we need to check
```
    if (currentRenderer == frameView->layoutContext().subtreeLayoutRoot())
        break;
```
 like what we do at the end of this for-loop? This `continue` will skip this check.

shallawa replied:
If `currentRenderer->isSVGRenderer()` but it corresponds to a `<use>` element and the position of one of the elements in the shadow tree was changed, will that set `requireFullLayout` eventually or not?

mattwoodrow replied:
The above refactoring means that we now hit the `subtreeLayoutRoot` check for all types. This wasn't strictly necessary, but can be a nice early exit.

Changing content inside the `<use>` should set layout flags on the ancestor that stops this fast path from being taken.

shallawa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3170,6 +3170,14 @@ bool Document::updateLayoutIfDimensionsOutOfDate(Element& element, OptionSet<Dim
         // Check our containing block chain. If anything in the chain needs a layout, then require a full layout.
         for (CheckedPtr currentRenderer = renderer; currentRenderer && !currentRenderer->isRenderView(); currentRenderer = currentRenderer->container()) {
 
+            if (currentRenderer->isSVGRenderer()) {
```
Instead of checking every renderer in the block chain if it `isSVGRenderer()`, I think this check can be moved inside the if-statement below:
```
    if (!currentRendererBox) {
        if (currentRenderer->isSVGRenderer()) {
            ...
        } else {
            requireFullLayout = true;
            break;
        }
    }
```

mattwoodrow replied:
This doesn't quite work unfortunately, since some SVG renderers (LegacyRenderSVGRoot) is also a box, and we want to use the SVG behaviour.

shallawa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3170,49 +3170,59 @@ bool Document::updateLayoutIfDimensionsOutOfDate(Element& element, OptionSet<Dim
+                    requireFullLayout = true;
+                    break;
+                }
+
```
An extra line here.

shallawa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3170,49 +3170,59 @@ bool Document::updateLayoutIfDimensionsOutOfDate(Element& element, OptionSet<Dim
+                    requireFullLayout = true;
+                    break;
+                }
+            } else if (!currentRenderer->isSVGRenderer()) {
```
I would suggest the following refractor:
```
    if (!currentRenderer->isSVGRenderer()) {
        CheckedPtr currentRendererBox = dynamicDowncast<RenderBox>(*currentRenderer);
        if (!currentRendererBox) {
            requireFullLayout = true;
            break;
        }
        // A box which is not an SVG
        ...
    }
    if (currentRenderer == frameView->layoutContext().subtreeLayoutRoot())
        break;
```

mattwoodrow committed:
Don't require a full layout to query size of SVG renderers.
https://bugs.webkit.org/show_bug.cgi?id=291185
<rdar://148727795>

Reviewed by Said Abou-Hallawa.

For SVG content, as long as the target doesn't need any form of layout, and
ancestors don't need self or positioned movement layout, then we should be able
to use the existing layout values.

It also updates clientLeft/Top to use this path (like width/height already do),
and adds a few missing ayoutOptions::IgnorePendingStylesheets flags.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetWidth):
(WebCore::Element::offsetHeight):
(WebCore::Element::clientLeft):
(WebCore::Element::clientTop):
(WebCore::Element::clientWidth):
(WebCore::Element::clientHeight):
(WebCore::Element::scrollWidth):
(WebCore::Element::scrollHeight):

Canonical link: https://commits.webkit.org/293444@main
